proposed/swift-interop.md

```diff
@@ -118,7 +118,13 @@ When calling a function that returns an opaque struct, the Swift ABI always requ
 
-For frozen structs and enums, Swift has a complicated lowering process where the struct or enum type's layout are recursively flattened to a sequence of primitives. If this sequence is length 4 or less, the values of this type are split into the elements of this sequence for parameter passing instead of passing the struct as a whole. Structs and enums that cannot be broken down in this way are passed by-reference to their specified frozen layout. Due to high implementation cost in the RyuJIT, in particular in the `UnmanagedCallersOnly` scenario, we should implement this first pass of lowering in the projection layer; the only types allowed for `CallConvSwift` calling convention in method or function pointer signatures are primitives, our special Swift register types, and pointer types. For reference, this lowering pass is done in the Swift compiler when lowering from Swift IL to LLVM IR. This design decision reinforces our direction of having the Runtime layer of Swift interop support similar features as the LLVM IR representation of Swift.
+For frozen structs and enums, Swift has a complicated lowering process where the struct or enum type's layout are recursively flattened to a sequence of primitives. If this sequence is length 4 or less, the values of this type are split into the elements of this sequence for parameter passing instead of passing the struct as a whole. Structs and enums that cannot be broken down in this way are passed by-reference to their specified frozen layout. When a frozen struct or enum with a valid primitive sequence of 4 elements or less is returned from a function, it is returned if it were a structure of the elements of the primitive sequence. Due to high implementation cost in the RyuJIT, in particular in the `UnmanagedCallersOnly` scenario, we should implement this first pass of lowering in the projection layer. The only types allowed for `CallConvSwift` calling convention in method or function pointer parameters are primitives, our special Swift register types, and pointer types. In return types, we will also allow structure types to support returning the primitive type sequences correctly. For reference, this lowering pass is done in the Swift compiler when lowering from Swift IL to LLVM IR. This design decision reinforces our direction of having the Runtime layer of Swift interop support similar features as the LLVM IR representation of Swift.
+
+##### SIMD Types
+
+We will pass the `System.Runtime.Intrinsics.VectorX<T>` types in SIMD registers as we do with the managed calling convention. We will treat the `Vector2/3/4` types as non-SIMD types (and block their usage directly as parameters in the `CallConvSwift` signature as is the case with other structs).
```

Comment:
```suggestion
We will pass the `System.Runtime.Intrinsics.VectorX<T>` types in SIMD registers. We will treat the `Vector2/3/4` types as non-SIMD types (and block their usage directly as parameters in the `CallConvSwift` signature as is the case with other structs).
```

I don't think the managed convention does this on all platforms (today).

Do we allow interop with these types in other interop scenarios? It sounds like it is going to add https://github.com/dotnet/runtime/issues/8300 + https://github.com/dotnet/runtime/issues/9578 as part of the work.

Reply:
I believe we have support on ARM64 due to HFA/HVA support. If I'm wrong, then yes this would add in those two issues as part of this work.

Reply:
Yes, on ARM64 we support it, but not on x64.

Is the SIMD interop important enough to warrant implementing it for x64? Those two issues on their own are large work items.

Reply:
It looks like macOS x64 is still going to be widely supported when .NET 9 releases, so it depends on if the libraries we want to support are high enough priority. For example, the [Accelerate framework](https://developer.apple.com/documentation/accelerate) has many APIs that take the SIMD types.

@kotlarmilos what are the Apple libraries that we're targeting for .NET 9?

Reply:
Why would `Vector2/3/4` be blocked? They have always been supported for interop and have been treated the equivalent of user-defined structs containing 2, 3, or 4 `float` fields (which is exactly how they are defined).

`Vector64/128/256/512<T>` and `Vector<T>` are all blocked from interop. The former set because Windows doesn't correctly handle SIMD returns today (this isn't vectorcall, but rather missing handling for the default x64 calling convention) and the latter because it doesn't make sense from an interop perspective today.

-----

> Yes, on ARM64 we support it, but not on x64.

This should exist for Unix already as well and only be missing for Windows x64, since that doesn't pass vectors differently (only returns them differently). `__vectorcall` would be required for Windows x64 HFA/HVA support and is still desirable long term so that we better optimize such perf critical functions; it just hasn't bubbled up in priority yet.

Reply:
`Vector2/3/4` would be blocked because they're already not handled as SIMD for interop and are allowed in P/Invokes (we have many tests validating that we don't break this behavior).

Swift on Windows is not something we're looking at supporting in the .NET 9 time frame, so that might help reduce cost here.

Reply:
I don't understand why this means they'd be blocked? Presumably they should work and be treated as any other user defined struct with 2-4 float fields, no?

Reply:
Yes, and we block any structs as parameters because we are requiring the "primitive type sequence" lowering to be done at the projection level. As a result, treating Vector2/3/4 as a regular struct means that they're blocked.

Reply:
> what are the Apple libraries that we're targeting for .NET 9?

We should start from https://github.com/dotnet/runtime/issues/95636. I don't expect any SIMD intrinsics in these libraries.

Reply:
> > Yes, on ARM64 we support it, but not on x64.
> 
> This should exist for Unix already as well and only be missing for Windows x64, since that doesn't pass vectors differently (only returns them differently). `__vectorcall` would be required for Windows x64 HFA/HVA support and is still desirable long term so that we better optimize such perf critical functions; it just hasn't bubbled up in priority yet.

I don't see us passing `VectorX` in SIMD registers on either SysV or win-x64. But you're right that we do support passing structs classified as such in SIMD registers if the SysV classifier says so, so at least there is some support in the backend for this. We wouldn't be able to use the SysV classifier here, however.

It helps that we don't have to consider Windows, and given was @kotlarmilos said it sounds like we can consider the full x64 support to be a stretch goal for now.

Reply:
(FWIW, I would still suggest incorporating my suggestion, since we don't pass VectorX types in SIMD registers across the board.)

Reply:
Implementing the primitive-type-sequence lowering in the JIT was deemed too expensive (see the comments on #310 for some comments from the JIT team).

I agree that I'm not a fan of pushing these details up to the projection layer. However, since our consumers are expected to use the Swift interop projection tooling to interact with Swift and not use the callconv handling directly, it was deemed acceptable.

We expect all Swift interop users to use Swift as the source-of-truth. Manually written "blittable P/Invokes" that are directly used like with our C interop story with a nice API shape as a C# source-of-truth are out-of-scope.

Reply:
I think we should re-evaluate the decision now that we understand the problem space in more detail. The split that is currently proposed does not feel right.

The JIT lowering does not need to produce best code quality. Does it make the problem easier?

Reply:
From @jakobbotsch below:

> I don't think this is similar to SysV. In the SysV classifier, there is just one struct layout and the classifier splits it out over a set of registers or passes the entire thing on stack. It's not hard to represent as a first-class ABI constraint in the JIT.
>
> I think this is more like ARM32. In ARM32, you can have single arguments that are passed partly in registers and partly on stack. The JIT has special support for it (FEATURE_ARG_SPLIT) and a special node PUTARG_SPLIT that is only used on ARM32 (and win-arm64 which has this peculiar case in an edge case around varargs too).
>
> Swift takes this to the next level. In Swift, you can have a single argument that interacts arbitrarily with the underlying calling convention. For example, a single struct argument can simultaneously have parts of it that are passed as implicit byrefs (copied to caller stack and a pointer passed), other parts of it that get passed in registers, other parts that are on stack, other parts that are HFAs.... There really is no limit to how many ABI constraints result from a single argument. That's what makes this harder to represent in a first-class way than any other ABI (especially in the UCO case).

Based on what we've learned (in particular that the type-sequence lowering goes down to primitives), I don't think the situation is as dire. I think the following are the basic pieces of how a frozen struct could be passed:

If the primitive type sequence is longer than 4 primitives:

- The whole struct is passed as implicit byref.

If the primitive type sequence is less than 4 primitives, each element of the sequence can be passed one of the following ways:

- In a general purpose register
- In a SIMD register
- On the stack (once registers are exhausted)

Because Swift is not supported on x86 Windows, we don't need to worry about the implicit byref case as no primitive is passed as an implicit byref on 64-bit platforms (to my knowledge).

We don't need to worry about the HFA case as the struct will already be broken down to its constituent float/double primitives.

@jakobbotsch does that simplify the scenario enough to make this implementable in the JIT?

Reply:
I share the concerns raised by others around exposing the detail to users. Just to be clear, the original discussion was around whether we could move some of this lowering from the JIT into the VM. My suggestion was that we would have the VM handle the lowering by creating IL stubs, in the same way that is done today if marshalling is needed. It would be transparent to the user and the user could have signatures with (frozen) structs and tuples.

The proposed design was already handling tuples in the projection layer, so it did not seem like a reach to handle frozen structs in the same way.

As I understand it the hard case with doing this in the VM is again the UCO case. How hard is it to create wrapper stubs in the VM that can handle the case even for UCO? Even if it is hard to do today, it seems like work that is similar to work we are going to end up needing for async2 efforts anyway. So if we can do that work today and avoid the complication in the JIT then I think this is way better. 

> The JIT lowering does not need to produce best code quality. Does it make the problem easier?

For calls I think we can handle the transformation in the JIT. We can split arguments as needed and represent them as the simpler ones with their normal ABI constraints (implicit byrefs, HFAs, registers, stack...). It probably requires some work around how we represent ABI constraints (today all arguments have associated metadata types, which they wouldn't have here), but this seems like "just work".

For UCO the only reasonable way to handle it is by spilling these to stack immediately, which as you say will result in terrible CQ. That's the case I'm worried about, since it requires making the backend work with a single parameter that has arbitrarily complicated ABI constraints associated with it. It's not impossible -- but I think doing this lowering in another place, such as the VM, will both be simpler and result in better codegen.

Reply:
> How hard is it to create wrapper stubs in the VM that can handle the case even for UCO? 

Very complicated. Impacts type system, diagnostics, reflection, ... .

> For UCO the only reasonable way to handle it is by spilling these to stack immediately, which as you say will result in terrible CQ.

If we went with wrapper stubs generated by VM, I expect that the CQ would be even worse.

Reply:
> From @jakobbotsch below:
> 
> > I don't think this is similar to SysV. In the SysV classifier, there is just one struct layout and the classifier splits it out over a set of registers or passes the entire thing on stack. It's not hard to represent as a first-class ABI constraint in the JIT.
> > I think this is more like ARM32. In ARM32, you can have single arguments that are passed partly in registers and partly on stack. The JIT has special support for it (FEATURE_ARG_SPLIT) and a special node PUTARG_SPLIT that is only used on ARM32 (and win-arm64 which has this peculiar case in an edge case around varargs too).
> > Swift takes this to the next level. In Swift, you can have a single argument that interacts arbitrarily with the underlying calling convention. For example, a single struct argument can simultaneously have parts of it that are passed as implicit byrefs (copied to caller stack and a pointer passed), other parts of it that get passed in registers, other parts that are on stack, other parts that are HFAs.... There really is no limit to how many ABI constraints result from a single argument. That's what makes this harder to represent in a first-class way than any other ABI (especially in the UCO case).
> 
> Based on what we've learned (in particular that the type-sequence lowering goes down to primitives), I don't think the situation is as dire. I think the following are the basic pieces of how a frozen struct could be passed:
> 
> If the primitive type sequence is longer than 4 primitives:
> 
> * The whole struct is passed as implicit byref.
> 
> If the primitive type sequence is less than 4 primitives, each element of the sequence can be passed one of the following ways:
> 
> * In a general purpose register
> * In a SIMD register
> * On the stack (once registers are exhausted)
> 
> Because Swift is not supported on x86 Windows, we don't need to worry about the implicit byref case as no primitive is passed as an implicit byref on 64-bit platforms (to my knowledge).
> 
> We don't need to worry about the HFA case as the struct will already be broken down to its constituent float/double primitives.
> 
> @jakobbotsch does that simplify the scenario enough to make this implementable in the JIT?

It seems workable, but it sounds like it is based on the assumption that we do not need to handle tuples in the JIT. Are you suggesting we would allow frozen structs but not tuples in signatures?

Reply:
> > How hard is it to create wrapper stubs in the VM that can handle the case even for UCO?
> 
> Very complicated. Impacts type system, diagnostics, reflection, ... .

Do you think the work will be avoided in the long run, or will it just be pushed into the future? Again I'm thinking about async2.

> > For UCO the only reasonable way to handle it is by spilling these to stack immediately, which as you say will result in terrible CQ.
> 
> If we went with wrapper stubs generated by VM, I expect that the CQ would be even worse on average.

I would not expect that. The JIT will be able to handle things efficiently through natural inlining and promotion of the constituent fields. We will have proper liveness that will allow the JIT to avoid the unnecessary spilling.

Reply:
Note that UCOs always need to move arguments to callee saved registers or spill them to stack as the very first thing to do `JIT_ReversePInvokeEnter` call.

Reply:
> Again I'm thinking about async2.

I think that the design for async2 is up in the air. It is more likely than not that the final async2 design is going to be quite different from what we have in the experiment prototype. I would not make any near term decisions based on async2.

> The JIT will be able to handle things efficiently through natural inlining and promotion of the constituent fields.

I do not think that the JIT will be able to eliminate the call overhead (fully inline the actual method) in average case. You can do a lot of spilling for the cost of a single call.

Reply:
We used to have the wrappers for UCOs on x86 and worked hard to get rid them.

Reply:
> I do not think that the JIT will be able to eliminate the call overhead (fully inline the actual method) in average case. You can do a lot of spilling for the cost of a single call.

It does not seem hard to make these particular calls aggressively inlined if we think that's beneficial.

Regardless, I'll trust your intuition about it. Let's try to see what will be needed on the JIT side to make the spilling approach work once we get to that point.

@jkoritzinsky I'm still curious about the above. Can we assume that the JIT does not need to handle tuples?

Reply:
My biggest concern here is that ABI is an extremely complex space and interop is one of those spaces where users want both simplicity and reduced overhead, especially when generating larger binding libraries.

Apple has also notably broken ABI in the past or deviated conventions from the norm on new platforms and so it is entirely possible some new platform comes on and now every single bit of `ObjC`/`Swift` interop code is DoA.

I think it is ultimately much better (even if its not what is done for the initial release due to timing constraints or w/e) that we have this support in the runtime as a detail of the `CallConv` support and that users are ultimately able to write a `delegate* unmanaged[CallConvSwift]<T, U, V>` that mirrors the underlying ObjC/Swift signature that would be exposed to C/C++ using the official Swift tooling.

Reply:
> It does not seem hard to make these particular calls aggressively inlined if we think that's beneficial.

These calls are typically going to have try/catch block in them to convert the .NET exception into switft error. You would have to implement inlining of methods with exception handling to make this work...

Reply:
That is to say, a user should be able to export Swift bindings using official Apple tooling and then use another existing tool, such as ClangSharp, CppAst, etc; which can generate blittable P/Invoke bindings from a C header and expect it to work.

If we can't achieve that, I expect we will have a lot of downstream pain/headaches from the community, especially as it gets into more complex bindings and libraries.

Reply:
> These calls are typically going to have try/catch block in them to convert the .NET exception into switft error. You would have to implement inlining of methods with exception handling to make this work...

Don't tempt me :-) (Note that this is actually part of our .NET 9 plan, and I also think it would be much more likely we end up with this support than appetite for improving the UCO Swift case in the future.)

Reply:
Re handling tuples: I think we can still handle tuples at the projection layer since the splitting of a tuple into separate arguments is done at the SIL layer and is very straightforward (it doesn't have nearly the same complexity as the "primitive type sequence" lowering) especially if JIT implementation cost for tuples would be too much.

Additionally, the primitive type sequence lowering happens after the tuple lowering (so each tuple element can be lowered to a sequence of up to 4 primitives), so handling tuples in the projection layer doesn't interfere with the primitive sequence handling.